### PR TITLE
Implement PlaceholderAPI support

### PR DESCRIPTION
**Pull request by PaulBGD**
Let me know if any of this needs changes.

The only thing off the top of my head that might not be perfect is that I removed the deepClone of the packet, so that commit can be dropped as needed. However, the performance improvement is significant (the cloning of the packet is up to 20-30% of a tick on a server with a lot of players/holograms.)

I do notice you have a v3 branch, I can update this for that if you'd prefer not to have this merged into the current version.

### CODE REVIEW

File: Plugin/src/main/java/com/gmail/filoghost/holographicdisplays/bridge/protocollib/current/ProtocolLibHookImpl.java [hunk 1]
```diff
@@ -57,8 +59,18 @@
 	private NMSManager nmsManager;
 	private PacketHelper packetHelper;
 	private MetadataHelper metadataHelper;
-	
-	
+	private final boolean useOptional;
+
+	public ProtocolLibHookImpl() {
+		String version = Bukkit.getServer().getClass().getPackage().getName().replace(".", ",").split(",")[3];
```
Comment (Andre601) [inline Plugin/src/main/java/com/gmail/filoghost/holographicdisplays/bridge/protocollib/current/ProtocolLibHookImpl.java:65]: Why first replace `.` with `,` and then split on it?
Could you not directly split at `.` by using `\\.`?

**Comment by Andre601** [comment]
Does this also update placeholders when needed?
Since some may have dynamic values that change with time would it be kinda sad to not have HD update those placeholders in a specific interval.

File: Plugin/src/main/java/com/gmail/filoghost/holographicdisplays/bridge/protocollib/current/ProtocolLibHookImpl.java (hunk 1 above)
Comment (PaulBGD) [reply Plugin/src/main/java/com/gmail/filoghost/holographicdisplays/bridge/protocollib/current/ProtocolLibHookImpl.java:65]: Honestly just reused code from here: https://github.com/Niall7459/HolographicExtension/blob/f25850f25ec91ac2dee9bee5f4314312713eb0f3/src/main/java/net/kitesoftware/holograms/listener/PacketPlaceholderListener.java#L40

(which is also GPL 3)

**Review by PaulBGD — COMMENTED** [review]
> Does this also update placeholders when needed?
> Since some may have dynamic values that change with time would it be kinda sad to not have HD update those placeholders in a specific interval.

Currently? No. I think that's a feature that would need further integration, and could be covered in another PR.

**Comment by Andre601** [comment]
HD already has something in place to update lines in intervals. As an example the `&u` rainbow colour or the custom animated text option.
So perhaps there could be a way to make HD update the line, if it contains a valid Placeholder or something...

**Comment by PaulBGD** [comment]
Imo I think that's more out of scope, as that's something that could be implemented outside of PAPI support.

**Comment by filoghost** [comment]
I'm implementing this changes in the v3 branch. Unfortunately cloning packets is needed, otherwise you're modifying a single packet for all players, and they would not be able to see different values.

Normal placeholders have changed a lot: i don't remember how it was previously, but the lines are updated only when necessary: if a placeholder updates but doesn't change values nothing is sent to the player. I will first implement it like yours and think about "auto refresh" later.

**Pull request closed without merging**